apps/wallet/src/shared/qredo-api.ts

```diff
@@ -0,0 +1,155 @@
+        null;
+
+    constructor(
+        qredoID: string,
+        baseURL: string,
+        options: {
+            accessToken?: string;
+            accessTokenRenewalFN?: AccessTokenRenewalFunction;
+        } = {}
+    ) {
+        this.qredoID = qredoID;
+        this.baseURL = baseURL + (baseURL.endsWith('/') ? '' : '/');
+        this.#accessToken = options.accessToken || null;
+        this.#renewAccessTokenFN = options.accessTokenRenewalFN || null;
+    }
+
+    public set accessToken(accessToken: string) {
+        this.#accessToken = accessToken;
+    }
+
+    public get accessToken() {
+        return this.#accessToken || '';
+    }
+
+    public createAccessToken({
+        refreshToken,
+        grantType = 'refresh_token',
+    }: AccessTokenParams): Promise<AccessTokenResponse> {
+        const params = new FormData();
+        params.append('refresh_token', refreshToken);
+        if (grantType) {
+            params.append('grant_type', grantType);
+        }
+        return this.#request(`${this.baseURL}connect/sui/token`, {
+            method: 'post',
+            body: params,
+        });
+    }
+
+    public getWallets({
+        filters,
+    }: GetWalletsParams = {}): Promise<GetWalletsResponse> {
+        const searchParams = new URLSearchParams();
+        if (filters?.address) {
+            searchParams.append('address', filters.address);
+        }
+        const searchQuery = searchParams.toString();
+        return this.#request(
+            `${this.baseURL}connect/sui/wallets${
+                searchQuery ? `?${searchQuery}` : ''
+            }`
+        );
+    }
+
+    #request = async (...params: Parameters<typeof fetch>) => {
+        let tries = 0;
+        while (tries++ <= 1) {
+            // TODO: add monitoring?
+            const response = await fetch(params[0], {
+                ...params[1],
+                headers: {
+                    ...params[1]?.headers,
+                    Authorization: `Bearer ${this.#accessToken}`,
+                },
+            });
+            const dataJson = await response.json();
+            if (response.ok) {
+                return dataJson;
+            }
+            if (response.status === 401) {
+                if (this.#renewAccessTokenFN && tries === 1) {
+                    if (this.#accessTokenRenewInProgress) {
+                        await this.#accessTokenRenewInProgress;
+                    } else {
+                        this.#accessTokenRenewInProgress =
+                            this.#renewAccessTokenFN(this.qredoID).finally(
+                                () => (this.#accessTokenRenewInProgress = null)
+                            );
+                        this.#accessToken = await this
+                            .#accessTokenRenewInProgress;
+                    }
+                    if (this.#accessToken) {
+                        continue;
```

Comment:
this control flow logic is a tad hard to follow (especially with all the nesting), I wonder if there's a simpler way to re-write this retry / refresh logic?